In [2]:
import matplotlib.pyplot as plt
%matplotlib inline
import numpy as np
import pandas as pd

# Q1>   
Supervised machine learning algorithm for task 1 - Regression for task 2 - Decision tree

In [33]:
# Loading the  data
abalone_data = pd.read_csv('abalone.csv')
abalone_data.columns
abalone_data.describe()
abalone_data.dtypes
abalone_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4177 entries, 0 to 4176
Data columns (total 9 columns):
Sex               4177 non-null object
Length            4177 non-null float64
Diameter          4177 non-null float64
Height            4177 non-null float64
Whole weight      4177 non-null float64
Shucked weight    4177 non-null float64
Viscera weight    4177 non-null float64
Shell weight      4177 non-null float64
Rings             4177 non-null int64
dtypes: float64(7), int64(1), object(1)
memory usage: 293.8+ KB


In [ ]:
np.corrcoef(abalone_data['Rings'] , abalone_data['Shell weight'])

# Question a..
No NULL values

In [113]:
abalone_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4177 entries, 0 to 4176
Data columns (total 9 columns):
Sex               4177 non-null category
Length            4177 non-null float64
Diameter          4177 non-null float64
Height            4177 non-null float64
Whole weight      4177 non-null float64
Shucked weight    4177 non-null float64
Viscera weight    4177 non-null float64
Shell weight      4177 non-null float64
Rings             4177 non-null int64
dtypes: category(1), float64(7), int64(1)
memory usage: 265.3 KB


catergorical variable - SEX

In [31]:
abalone_data.corr()

Length  Diameter    Height  Whole weight  Shucked weight  \
Length          1.000000  0.986812  0.827554      0.925261        0.897914   
Diameter        0.986812  1.000000  0.833684      0.925452        0.893162   
Height          0.827554  0.833684  1.000000      0.819221        0.774972   
Whole weight    0.925261  0.925452  0.819221      1.000000        0.969405   
Shucked weight  0.897914  0.893162  0.774972      0.969405        1.000000   
Viscera weight  0.903018  0.899724  0.798319      0.966375        0.931961   
Shell weight    0.897706  0.905330  0.817338      0.955355        0.882617   
Rings           0.556720  0.574660  0.557467      0.540390        0.420884   

                Viscera weight  Shell weight     Rings  
Length                0.903018      0.897706  0.556720  
Diameter              0.899724      0.905330  0.574660  
Height                0.798319      0.817338  0.557467  
Whole weight          0.966375      0.955355  0.540390  
Shucked weight        0.931961      0.882617  0.420884  
Viscera weight        1.000000      0.907656  0.503819  
Shell weight          0.907656      1.000000  0.627574  
Rings                 0.503819      0.627574  1.000000

In [32]:
import statsmodels.api as sms

In [8]:
price_model = sms.OLS(abalone_data['Rings'],
                      abalone_data['Shell weight'])

In [9]:
results = price_model.fit() 

In [10]:
results.summary() 

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Rings   R-squared:                       0.845
Model:                            OLS   Adj. R-squared:                  0.845
Method:                 Least Squares   F-statistic:                 2.280e+04
Date:                Tue, 10 Sep 2019   Prob (F-statistic):               0.00
Time:                        09:01:01   Log-Likelihood:                -11830.
No. Observations:                4177   AIC:                         2.366e+04
Df Residuals:                    4176   BIC:                         2.367e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Shell weight    34.7332      0.230    150.986      0.000      34.282      35.184
==============================================================================
Omnibus:                      208.414   Durbin-Watson:                   0.591
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              291.907
Skew:                          -0.465   Prob(JB):                     4.10e-64
Kurtosis:                       3.901   Cond. No.                         1.00
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

###### The above model indicates the linear function

Rings = 34.7332*Shell weight

In [11]:
predictors = np.array([abalone_data['Shell weight'],
          [1]*abalone_data['Shell weight'].shape[0]]).T

In [13]:
price_model_w_b0 = sms.OLS(abalone_data['Rings'],
                      predictors)

In [14]:
res1  =price_model_w_b0.fit()

In [15]:
res1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Rings   R-squared:                       0.394
Model:                            OLS   Adj. R-squared:                  0.394
Method:                 Least Squares   F-statistic:                     2713.
Date:                Tue, 10 Sep 2019   Prob (F-statistic):               0.00
Time:                        09:06:48   Log-Likelihood:                -9770.8
No. Observations:                4177   AIC:                         1.955e+04
Df Residuals:                    4175   BIC:                         1.956e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1            14.5357      0.279     52.084      0.000      13.989      15.083
const          6.4621      0.077     83.764      0.000       6.311       6.613
==============================================================================
Omnibus:                     1213.646   Durbin-Watson:                   1.016
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3601.423
Skew:                           1.503   Prob(JB):                         0.00
Kurtosis:                       6.414   Cond. No.                         7.60
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

# MODEL

In [34]:
features = abalone_data.columns.drop(['Rings'])

In [35]:
from sklearn.linear_model import LinearRegression

In [36]:
abalone_data.columns

Index(['Sex', 'Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weight',
       'Viscera weight', 'Shell weight', 'Rings'],
      dtype='object')

In [37]:
abalone_data['Sex'] = abalone_data['Sex'].astype('category')

In [94]:
abalone_data.columns


Index(['Sex', 'Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weight',
       'Viscera weight', 'Shell weight', 'Rings'],
      dtype='object')

In [95]:
cat_vars=['Sex']
abalone_data_cat = pd.get_dummies(abalone_data,columns=cat_vars)

In [96]:
abalone_data_cat.columns

Index(['Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weight',
       'Viscera weight', 'Shell weight', 'Rings', 'Sex_F', 'Sex_I', 'Sex_M'],
      dtype='object')

In [100]:
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
min_max_scaler = preprocessing.MinMaxScaler()
abalone_data_cat.head()
# abalone_data_train =  abalone_data_cat.drop('Rings',axis=1)
# abalone_data_train.head()
#abalone_data_minmax = min_max_scaler.fit_transform(abalone_data['Length']).reshape(-1, 1) 
# scaler = MinMaxScaler()
# scaler.fit(abalone_data_train)
# abalone_data_train = scaler.transform(abalone_data_train)


Length  Diameter  Height  Whole weight  Shucked weight  Viscera weight  \
0   0.455     0.365   0.095        0.5140          0.2245          0.1010   
1   0.350     0.265   0.090        0.2255          0.0995          0.0485   
2   0.530     0.420   0.135        0.6770          0.2565          0.1415   
3   0.440     0.365   0.125        0.5160          0.2155          0.1140   
4   0.330     0.255   0.080        0.2050          0.0895          0.0395   

   Shell weight  Rings  Sex_F  Sex_I  Sex_M  
0         0.150     15      0      0      1  
1         0.070      7      0      0      1  
2         0.210      9      1      0      0  
3         0.155     10      0      0      1  
4         0.055      7      0      1      0

One hot encoding is done

In [103]:
features = ['Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weight',
       'Viscera weight', 'Shell weight', 'Sex_I', 'Sex_M','Sex_F']
target = ['Rings']
model4 = LinearRegression()


model4.fit(abalone_data_cat[features],abalone_data_cat[target])
model4.score(abalone_data_cat[features],abalone_data_cat[target])

0.5372952716149669

In [104]:
abalone_data['Sex'].value_counts()

M    1528
I    1342
F    1307
Name: Sex, dtype: int64

Yes it is balanced data

In [105]:
from sklearn.model_selection import train_test_split

train, test = train_test_split(abalone_data_cat, test_size=0.2)

In [107]:
train.shape

(3341, 11)

In [108]:
test.shape

(836, 11)

In [109]:
features = ['Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weight',
       'Viscera weight', 'Shell weight', 'Sex_I', 'Sex_M','Sex_F']
target = ['Rings']
model4 = LinearRegression()


model4.fit(train[features],train[target])
model4.score(test[features],test[target])

0.5438504079383244

Test accuracy : 0.5438504079383244
Train accuracy : 0.5372952716149669

Some more data processing is required .
Accuracy on both train and test data are almost same


In [114]:
from sklearn.ensemble import RandomForestClassifier,AdaBoostClassifier,GradientBoostingClassifier

In [119]:
features = ['Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weight',
       'Viscera weight', 'Shell weight', 'Sex_I', 'Sex_M','Sex_F']
target = ['Rings']
model4 = GradientBoostingClassifier()


model4.fit(train[features],train[target])
model4.score(test[features],test[target])

D:\Anaconda\lib\site-packages\sklearn\utils\validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.2679425837320574

above is using   ensemble learner (Gradient boosting)